### PR TITLE
Collaborators with edit permissions can publish edits

### PR DESCRIPTION
**Pull request by Raemon**
This allows users with 'edit' permissions on post to edit and publish the entire post. 

Theoretically we might want to separate out multiple tiers of editing privileges, but basically Ruby's been getting a lot of complaints on intercom about people not being able to give coauthors permission to edit, and it seems like most people would prefer to be able to just give people full editing permissions if you're going to bother collaborating at all.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1202992167636772) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/posts/PostsEditForm.tsx [hunk 1]
```diff
@@ -58,11 +58,7 @@ const PostsEditForm = ({ documentId, classes }: {
 
   // If we only have read access to this post, but it's shared with us,
   // redirect to the collaborative editor.
-  if (document
-    && document.userId!==currentUser._id
-    && document.sharingSettings
-    && !userCanDo(currentUser, 'posts.edit.all')
-  ) {
+  if (document && !canUserEditPost(currentUser, document) && document.sharingSettings) {
```
Comment (Will-Howard) [inline packages/lesswrong/components/posts/PostsEditForm.tsx:61]: What is the check for `document.sharingSettings` achieving here?

**Review by Will-Howard — COMMENTED** [review]
Nice, I agree this is what the logic should be. I'll approve when the tests are fixed

File: packages/lesswrong/lib/collections/users/helpers.tsx [hunk 2]
```diff
@@ -56,6 +56,22 @@ export const userIsSharedOn = (currentUser: DbUser|UsersMinimumInfo|null, docume
   }
 }
 
+export const canUserEditPost = (user: DbUser|UsersCurrent|null, doc: PostsList|DbPost): boolean => {
+  if (!user) return false;
+
+  if (userOwns(user, doc)) return true
+
+  if (userCanDo(user, 'posts.edit.all'))
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/users/helpers.tsx:64]: This seems to be missing the `return true`?

File: packages/lesswrong/components/posts/PostsEditForm.tsx (hunk 1 above)
Comment (Raemon) [reply packages/lesswrong/components/posts/PostsEditForm.tsx:61]: So this predates me and I'm not 100% sure of the goals, but I think the idea is that you should only end up on the PostCollaborationEditor if the post actually has sharing settings.

File: packages/lesswrong/components/posts/PostsEditForm.tsx [hunk 3]
```diff
@@ -93,7 +89,7 @@ const PostsEditForm = ({ documentId, classes }: {
     return <Components.ErrorAccessDenied/>
   }
 
-  if (document?.fmCrosspost?.isCrosspost && !document?.fmCrosspost?.hostedHere) {
+  if (document.fmCrosspost?.isCrosspost && !document.fmCrosspost?.hostedHere) {
```
Comment (darkruby501) [inline packages/lesswrong/components/posts/PostsEditForm.tsx:92]: We should check that this plays nicely with the collab redirects

**Review by darkruby501 — CHANGES_REQUESTED** [review]

**Comment by Raemon** [comment]
Update:

I realized I can created a duplicate permission function. The posts/helpers.ts file had a `postCanEdit` function, similar to my new canUserEditPost function (which I had mistakenly put in users/helpers.ts).

I merged them into `posts/helpers.ts`, and I used the name `canUserEditPost` because Robert sold me on it being important for functions to have a naming schema that's harder to confuse with a variable. 

The downside here is that now this touches a lot more use-cases and probably warrants more thorough testing. 

Meanwhile I'm actually a bit confused about the test failures. I think probably I can figure it out when I sit down to look at them but it's not a trivial fix for me with my current background knowledge.

**Review by darkruby501 — APPROVED** [review]

**Pull request merged**